ProcessMaker/Jobs/DownloadSecurityLog.php

```diff
@@ -37,18 +50,161 @@ public function __construct(User $user, string $format, int $userId = null)
      * Execute the job.
      *
      * @return void
+     * @throw Exception
      */
     public function handle()
     {
-        //1. Get all data from security_logs table
-        //2. Create a file in specified format: csv or xml
-        //3. Zip this file
-        //4. Store in S3, with private visibility, with 24h of lifecycle
-        //5. Send email to user with the link or error message
-        if (mt_rand(1, 10) <= 8) {
-            event(new SecurityLogDownloadJobCompleted($this->user, true, __('Click on the link and download the file. This link will be available until midnight tonight.'), 'http://processmaker.com/?download=true'));
-        } else {
-            event (new SecurityLogDownloadJobCompleted($this->user, false, __('Sorry, it was not possible to generate the log file. Please contact the administrator.')));
+        if (!Media::s3IsReady()) {
+            event(new SecurityLogDownloadJobCompleted($this->user, false, __('Sorry, this feature requires the configured AWS S3 service.. Please contact the administrator.')));
+            return;
+        }
+        try {
+            $collection = $this->getCollection();
+            $contents = $this->generateContent($collection);
+            $expires = $this->getExpires();
+            $filename = $this->createTemporaryFilename();
+            $url = $this->createTemporaryUrl($filename, $contents, $expires);
+            $message = __('Click on the link and download the file. This link will be available until midnight tonight.');
+            event(new SecurityLogDownloadJobCompleted($this->user, true, $message, $url));
+        } catch (Exception $e) {
+            $message = __('Sorry, it was not possible to generate the log file. Please contact the administrator.');
+            event(new SecurityLogDownloadJobCompleted($this->user, false, $message));
+        }
+    }
+
+    /**
+     * Generate the content with the corresponding format
+     */
+    protected function generateContent(array $collection)
+    {
+        return $this->format === static::FORMAT_CSV ? $this->toCSV($collection) : $this->toXML($collection);
+    }
+
+    /**
+     * To CSV
+     */
+    protected function toCSV(array $collection)
+    {
+        $first = current($collection);
+        $headers = array_keys($first ?: []);
+        $content = '';
+        $line = [];
+        foreach ($headers as $key) {
+            $line[] = $key;
+        }
+        $content .= implode(static::CSV_SEPARATOR, $line);
+        foreach ($collection as $item) {
+            $line = [];
+            foreach ($headers as $key) {
+                if (is_object($item[$key])) {
+                    $line[] = json_encode($item[$key]);
+                } else {
+                    $line[] = isset($item[$key]) ? $item[$key] : null;
+                }
+            }
+            $content .= PHP_EOL . implode(static::CSV_SEPARATOR, $line);
```

Comment:
If the content is going to be very large, could you consider to stream the content instead of preparing all the result in memory?